src/utils/nps.ts

```diff
@@ -0,0 +1,85 @@
+
+const PROBABILITY = 0.15;
+const MIN_SESSION_COUNT = 10;
+
+const SURVEY_NAME = 'nps1';
+const SURVEY_URL = 'https://www.surveymonkey.com/r/vscodedockernpsinproduct';
+
+const SESSION_COUNT_KEY = `${SURVEY_NAME}/sessioncount`;
+const LAST_SESSION_DATE_KEY = `${SURVEY_NAME}/lastsessiondate`;
+const IS_CANDIDATE_KEY = `${SURVEY_NAME}/iscandidate`;
+
+export async function nps(globalState: Memento): Promise<void> {
+    try {
+        // If not English-language, don't ask
+        if (env.language !== 'en' && !env.language.startsWith('en-')) {
+            return;
+        }
+
+        let isCandidate: boolean | undefined = globalState.get(IS_CANDIDATE_KEY);
+
+        // If not a candidate, don't ask
+        if (isCandidate === false) {
+            return;
+        }
+
+        const date = new Date().toDateString();
+        const lastSessionDate = globalState.get(LAST_SESSION_DATE_KEY, new Date(0).toDateString());
+
+        // If this session is on same date as last session, don't count it
+        if (date === lastSessionDate) {
+            return;
+        }
+
+        // Count this session
+        const sessionCount = globalState.get(SESSION_COUNT_KEY, 0) + 1;
+        await globalState.update(LAST_SESSION_DATE_KEY, date);
+        await globalState.update(SESSION_COUNT_KEY, sessionCount);
+
+        // If under the MIN_SESSION_COUNT, don't ask
+        if (sessionCount < MIN_SESSION_COUNT) {
+            return;
+        }
+
+        // Decide if they are a candidate (if we previously decided they are and they did Remind Me Later, we will not do probability again)
+        // i.e. Probability only comes into play if isCandidate is undefined
+        // tslint:disable-next-line: insecure-random
+        isCandidate = isCandidate || Math.random() < PROBABILITY;
+        await globalState.update(IS_CANDIDATE_KEY, isCandidate);
+
+        // If not a candidate, don't ask
+        if (!isCandidate) {
+            return;
+        }
+
+        const take: MessageItem = { title: 'Take Survey' };
+        const remind: MessageItem = { title: 'Remind Me Later' };
+        const never: MessageItem = { title: 'Don\'t Show Again' };
+
+        // Prompt, treating hitting X as Remind Me Later
+        const result = (await window.showInformationMessage('Do you mind taking a quick feedback survey about the Docker Extension for VS Code?', take, remind, never)) || remind;
+
+        ext.reporter.sendTelemetryEvent('nps', { survey: SURVEY_NAME, response: result.title });
+
+        if (result === take) {
+            // If they hit Take, don't ask again (for this survey name), and open the survey
+            await globalState.update(IS_CANDIDATE_KEY, false);
+            await env.openExternal(Uri.parse(`${SURVEY_URL}?o=${encodeURIComponent(process.platform)}&m=${encodeURIComponent(env.machineId)}`));
+        } else if (result === remind) {
+            // If they hit the X or Remind Me Later, ask again in 3 sessions

```

Comment:
The session count monotonically increases, so once above 13 sessions, the user still has a chance of being prompted on each subsequent session.  Perhaps set the session count to `MIN_SESSION_COUNT - 3`?

Reply:
At session 10, we will decide a value for isCandidate, and save it (lines 56, 57). If it decides they are not a candidate they will never be prompted (line 31), otherwise, it goes through the take/remind/never flow. Take and never both result in isCandidate becoming false, so they will never be prompted. Remind me brings the session count down to 7 but leaves isCandidate true, so when it gets back up to session 10, we will prompt again.

Reply:
X is treated as remind later, so how could the session count ever get above 10?